File: problems/hard/bear_and_gene/bear_and_gene.c
#include <stdio.h>
#include <string.h>
#include <limits.h>

#define MAX_LEN 500001

/**
 * Finds the length of the smallest substring that can be replaced 
 * to make the gene steady.
 *
 * @param gene The input gene string.
 * @return The length of the smallest substring to replace.
 */
int steadyGene(char* gene) {    
    int n = strlen(gene);
    int required = n / 4;
    int count[128] = {0};

    for (int i = 0; i < n; i++) {
        count[(int)gene[i]]++;
    }

    if (count['A'] == required && count['C'] == required &&
        count['G'] == required && count['T'] == required) {
        return 0;
    }

    int minLen = n;
    int left = 0;
    for (int right = 0; right < n; right++) {
        count[(int)gene[right]]--;
        while (left < n && count['A'] <= required && count['C'] <= required &&
               count['G'] <= required && count['T'] <= required) {
            if (right - left + 1 < minLen) {
                minLen = right - left + 1;
            }
            count[(int)gene[left]]++;
            left++;
        }
    }
    return minLen;
}

int main() {
    // Example usage:
    char gene[MAX_LEN];
    scanf("%s", gene);
    int result = steadyGene(gene);
    printf("%d\n", result);
    return 0;
}

File: problems/hard/bear_and_gene/BearAndGene.java
package problems.hard.bear_and_gene;
import java.util.Scanner;

public class BearAndGene {

    public static int steadyGene(String gene) {
         int n = gene.length();
        int required = n / 4;
        int[] count = new int[128]; // For A, C, G, T
        for (char c : gene.toCharArray()) {
            count[c]++;
        }

        if (count['A'] == required && count['C'] == required &&
            count['G'] == required && count['T'] == required) {
            return 0;
        }

        int minLen = n, left = 0;
        for (int right = 0; right < n; right++) {
            count[gene.charAt(right)]--;
            while (left < n && count['A'] <= required && count['C'] <= required &&
                   count['G'] <= required && count['T'] <= required) {
                minLen = Math.min(minLen, right - left + 1);
                count[gene.charAt(left)]++;
                left++;
            }
        }
        return minLen;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter gene: ");
        String gene = sc.nextLine();
        System.out.println(steadyGene(gene));
        sc.close();
    }
}
